src/Microsoft.AspNet.Diagnostics/ErrorPageMiddleware.cs

```diff
@@ -182,31 +188,103 @@ private StackFrame StackFrame(Chunk line)
+            if (string.IsNullOrEmpty(file))
+            {
+                return GetStackFrame(
+                    string.IsNullOrEmpty(function) ? line.ToString() : function,
+                    file: string.Empty,
+                    lineNumber: 0,
+                    fileProvider: _options.FileProvider,
+                    sourceCodeLineCount: _options.SourceCodeLineCount);
+            }
+            else
+            {
+                return GetStackFrame(function, file, lineNumber, _options.FileProvider, _options.SourceCodeLineCount);
+            }
         }
 
-        private StackFrame LoadFrame(string function, string file, int lineNumber)
+        internal static StackFrame GetStackFrame(
+            string function,
+            string file,
+            int lineNumber,
+            IFileProvider fileProvider,
+            int sourceCodeLineCount)
         {
             var frame = new StackFrame { Function = function, File = file, Line = lineNumber };
+
+            if (string.IsNullOrEmpty(file))
+            {
+                return frame;
+            }
+
+            IEnumerable<string> lines = null;
             if (File.Exists(file))
             {
-                var code = File.ReadLines(file);
-                ReadFrameContent(frame, code, lineNumber, lineNumber);
+                lines = File.ReadLines(file);
             }
+            else
+            {
+                // Handle relative paths and embedded files
+                var fileInfo = fileProvider.GetFileInfo(file);
+                if (fileInfo.Exists)
+                {
+                    // ReadLines doesn't accept a stream. Use ReadLines as its more efficient
+                    // relative to reading lines via stream reader
+                    if (!string.IsNullOrEmpty(fileInfo.PhysicalPath))
+                    {
+                        lines = File.ReadLines(fileInfo.PhysicalPath);
+                    }
+                    else
+                    {
+                        lines = GetLines(fileInfo.CreateReadStream());
+                    }
+                }
+            }
+
+            if (lines != null)
+            {
+                GetErrorCodeBlock(frame, lines, lineNumber, sourceCodeLineCount);
+            }
+
             return frame;
         }
 
-        private void ReadFrameContent(StackFrame frame,
-                                      IEnumerable<string> code,
-                                      int startLineNumber,
-                                      int endLineNumber)
+        private static void GetErrorCodeBlock(
+            StackFrame frame,
+            IEnumerable<string> allLines,
+            int errorLineNumberInFile,
+            int sourceCodeLineCount)
+        {
+            // Get the line boundaries in the file to be read and read all these lines at once into an array.
+            // This will cause the backing stream to be closed.
```

Comment:
I modified the function signature to accept a IFileInfo instead...the behavior now should be similar to how it would be when we call `File.ReadLines` which internally opens the file stream and disposes it when the iterator internally is disposed which happens with the call to `ToArray`